risc0/zkp/rust/src/prove/merkle.rs

```diff
@@ -125,3 +125,229 @@ impl<H: Hal> MerkleTreeProver<H> {
+        queries: usize,
+    ) -> MerkleTreeProver<H> {
+        // Initialize a prover with leaves 0..size
+        let size: u32 = (rows * cols).try_into().unwrap();
+        let mut data = Vec::<Fp>::new();
+        for val in 0..size {
+            data.push(Fp::from(val));
+        }
+        let matrix = hal.copy_fp_from(data.as_slice());
+
+        MerkleTreeProver::new(hal, &matrix, rows, cols, queries)
+    }
+
+    fn bad_row_access<H: Hal, S: Sha>(sha: &S, hal: &H, rows: usize, cols: usize, queries: usize) {
+        let prover = init_prover(hal, rows, cols, queries);
+        let mut iop: WriteIOP<S> = WriteIOP::new(sha);
+        prover.prove(&mut iop, rows);
+    }
+
+    fn possibly_bad_verify<H: Hal, S: Sha>(
+        sha: &S,
+        hal: &H,
+        rows: usize,
+        cols: usize,
+        queries: usize,
+        bad_query: usize,
+        manipulate_proof: bool,
+    ) {
+        let prover = init_prover(hal, rows, cols, queries);
+
+        let mut iop: WriteIOP<S> = WriteIOP::new(sha);
+        prover.commit(hal, &mut iop);
+        for _query in 0..queries {
+            let r_idx = (iop.rng.next_u32() as usize) % rows;
+            let col = prover.prove(&mut iop, r_idx);
+            for c_idx in 0..cols {
+                assert_eq!(col[c_idx], Fp::from((r_idx + c_idx * rows) as u32));
+            }
+        }
+        {
+            if manipulate_proof {
+                let mut rng = rand::thread_rng();
+                let manip_idx = rng.gen::<usize>() % iop.proof.len();
+                iop.proof[manip_idx] ^= 1;
+            }
+            let mut r_iop = ReadIOP::new(sha, &iop.proof);
+            let verifier = MerkleTreeVerifier::new(&mut r_iop, rows, cols, queries);
+            assert_eq!(verifier.root(), prover.root());
+            for query in 0..queries {
+                let r_idx = (r_iop.next_u32() as usize) % rows;
+                if query == bad_query {
+                    let r_idx = (r_idx + 1) % rows;
+                    verifier.verify(&mut r_iop, r_idx);
+                }
+                let col = verifier.verify(&mut r_iop, r_idx);
+                for c_idx in 0..cols {
+                    assert_eq!(col[c_idx], Fp::from((r_idx + c_idx * rows) as u32));
+                }
+            }
+            r_iop.verify_complete();
+        }
+    }
+
+    fn randomize_sizes() -> (usize, usize, usize) {
+        // Chooses random values of `rows`, `cols`, and `queries` such that:
+        // `rows` is a power of 2
+        // `cols` & `queries` have a wide distribution but tend to take small values
+        let mut rng = rand::thread_rng();
+        let rows = 1 << (rng.gen::<u8>() % 10);
+        let cols = (rng.gen::<usize>() % (1 << (rng.gen::<u8>() % 10))) + 1;
+        let queries = (rng.gen::<usize>() % (1 << (rng.gen::<u8>() % 10))) + 1;
+        (rows, cols, queries)
+    }
+
+    #[test]
+    #[should_panic(expected = "assertion failed: idx < self.params.row_size")]
```

Comment:
This is probably fine for now but I think in the future we should avoid using panics and instead use `Result` to do error reporting. Panic won't work well in some environments since a panic causes the entire thread to be terminated.

Reply:
These tests would be cleaner if this was `Result`-based error reporting than panics anyway. I'm tracking down an intermittent failure in this test anyway -- perhaps the most straightforward approach will be to rewrite the tested code to use `Result`s.

Reply:
fyi I switched to `Result`-based errors on the verifier side, but not here on the prover side. We should definitely do it eventually, but there is more to be done on the prover side and IMO it's less important on the prover side.